contents/regex/Regex.md

```diff
@@ -0,0 +1,367 @@
+
+Imagine you're a witness of a robbery. How might you describe the robber to the police? You might say that the robber has a beard, brown hair and a facial scar. With this information, the police can round up everyone with beards, brown hair and scars as possible suspects. This is similar to how Regex works. You specify a search pattern using a string (e.g. It starts with 'C', ends with a `S`, and only consists of letters: `C[A-z]+S`) and regex searches a section of text for you, finding all the possible matches.
+
+## Why should I learn Regex?
+
+* **Regex is everywhere** - Regular Expressions are technically a language by themselves, but they are usually implemented within other languages. Almost all commonly used languages, like Java, C++ or Python supports Regex is some fashion (A full list is available [here](http://www.regular-expressions.info/tools.html)). In addition, many IDEs or text editors support Regex in 'Find and replace' operations, such as Eclipse, Visual Studio or Notepad++, allowing non-programmers to benefit from Regex.
+* **Regex is portable** - People often complain that switching to a new language usually means having to learn a new API and syntax. Regex however is implemented almost the same way on every single language, so you can just learn it once and use it everywhere.
+* **Regex is useful** - Although Regex doesn't have a specific use, it is a general use library with many different applications. It can be used automate many different common programming tasks such as validating user input, parsing data from text, and string manipulation, all of which are commonly done in any programming project.
+* **Regex is convenient** - "Regex offers a simple and painless way of dealing some with them" because not all string processing are regex-related.
+
+
+# Getting Started
+
+## Introduction
+
+The power of regex revolves around the idea that most data comes in a regular, predictable format (e.g. serial numbers, collected data, statistics). Some everyday examples are as follows:
+
+* Phone numbers: +65 65162727
+* Passwords: Contains lower/upper case, symbol…
+* IP Addresses: 127.0.0.1:8080
+* Date time: DD/MM/YY HH:MM
+
+Regex exploits these simlarities and provides a way of specifying patterns that can fit almost any use case. Some of the basic functionalities are as follows:
+
+Let's take a look at some simple examples without code. For this guide, you can use any text editor that supports Regex functionality (e.g. [Notepad++](https://notepad-plus-plus.org/download/v7.3.1.html), [Sublimetext](https://www.sublimetext.com/)).
+
+## Search Text
+
+The most basic use of Regex is searching a length of text. Say you have want to find all occurences of a word in a text. Simple, right? Just use the search function. However, what if you aren't looking for a specific string? For example:
+
+> * Find all phone numbers in a chat log
+> * Find all email addresses in a saved email
+> * Find all web URLs on a web page
```

Comment:
@damithc For this one I wanted to bring up examples in particular that can't be done with a simple Ctrl-F. Will change it to just going in depth into one example.